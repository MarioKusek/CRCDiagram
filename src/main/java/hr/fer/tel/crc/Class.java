package hr.fer.tel.crc;

import java.util.LinkedList;
import java.util.List;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class Class {

  private final String name;
  private String alias;
  private List<Responsibility> responsibilities;


  public Class(String name) {
    if(name == null)
      throw new NullPointerException("Name can not be null.");
    this.name = name;
    this.responsibilities = new LinkedList<>();
  }

  public List<Responsibility> getResponsibilities() {
    return responsibilities;
  }
}
