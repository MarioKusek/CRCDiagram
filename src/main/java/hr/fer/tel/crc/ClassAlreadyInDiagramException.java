package hr.fer.tel.crc;

public class ClassAlreadyInDiagramException extends RuntimeException {

  private ClassAlreadyInDiagramException(String message) {
    super(message);
  }

  public static ClassAlreadyInDiagramException withName(String className) {
    return new ClassAlreadyInDiagramException("Class with name " + className + " is already in diagram.");
  }

  public static ClassAlreadyInDiagramException withAlias(String name) {
    return new ClassAlreadyInDiagramException("Class with alias " + name + " is already in diagram.");
  }
}
