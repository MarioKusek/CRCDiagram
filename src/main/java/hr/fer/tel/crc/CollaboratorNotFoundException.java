package hr.fer.tel.crc;

public class CollaboratorNotFoundException extends RuntimeException {

  public CollaboratorNotFoundException(String className, String collaborator) {
    super("Class with name " + className + " has collaborator " + collaborator + " that can not be found.");
  }
}
