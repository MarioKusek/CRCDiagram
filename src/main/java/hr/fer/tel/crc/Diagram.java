package hr.fer.tel.crc;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Diagram {

  private List<Class> classes;
  private Map<String, Class> aliasses;

  public Diagram(List<Class> classes) {
    this.classes = classes;

    aliasses = new HashMap<>();
    classes.stream().forEach(c -> {
      if(aliasses.containsKey(c.getName()))
        throw new RuntimeException("Class with name " + c.getName() + " is already in diagram.");
      aliasses.put(c.getName(), c);

      if(aliasses.containsKey(c.getAlias()))
        throw new RuntimeException("Class with alias " + c.getName() + " is already in diagram.");
      if(c.getAlias() != null)
        aliasses.put(c.getAlias(), c);
    });

    checkCollaborators();
  }

  private void checkCollaborators() {
    classes.stream()
      .forEach(c -> {
        c.getResponsibilities().stream().forEach(r -> {
          String collaborator = r.getCollaborator();
          if(collaborator != null && getClassByKey(collaborator) == null)
            throw new RuntimeException("Class with name " + c.getName() + " has collaborator " + collaborator + " that can not be found.");
        });
      });
  }

  public Class getClassByKey(String key) {
    return aliasses.get(key);
  }

  public List<Class> getClasses() {
    return classes;
  }
}
