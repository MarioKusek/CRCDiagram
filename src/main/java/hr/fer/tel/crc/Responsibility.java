package hr.fer.tel.crc;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;

@RequiredArgsConstructor
@Getter
@Setter
public class Responsibility {
  private final String text;
  private String collaborator;
}
