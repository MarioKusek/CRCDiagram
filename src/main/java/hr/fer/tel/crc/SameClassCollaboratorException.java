package hr.fer.tel.crc;

public class SameClassCollaboratorException extends RuntimeException {

  public SameClassCollaboratorException(String className) {
    super("Class with name " + className + " must not have collaborator that is the same class.");
  }

}
