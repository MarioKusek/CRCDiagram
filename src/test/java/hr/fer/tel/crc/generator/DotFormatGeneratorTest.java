package hr.fer.tel.crc.generator;

import java.io.StringWriter;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import hr.fer.tel.crc.Class;
import hr.fer.tel.crc.Diagram;
import hr.fer.tel.crc.Responsibility;

class DotFormatGeneratorTest {

  private StringWriter writer = new StringWriter();
  private Diagram diagram;
  private DotGenerator generator;

  @Test
  void emptyDiagram() throws Exception {
    diagram = new Diagram(List.of());
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void oneClassWithoutResponsibilities() throws Exception {
    diagram = new Diagram(List.of(new Class("className1")));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void twoClassesWithoutResponsibilities() throws Exception {
    diagram = new Diagram(List.of(new Class("className1"), new Class("secondClassName")));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void oneClassWithResponsibility() throws Exception {
    diagram = new Diagram(List.of(
        Class.builder()
          .name("className1")
          .responsibility(new Responsibility("c1 resp1"))
          .build()));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void twoClassesWithResponsibilities() throws Exception {
    diagram = new Diagram(List.of(
      Class.builder()
        .name("className1")
        .responsibility(new Responsibility("c1 resp1"))
        .build(),
      Class.builder()
        .name("className2")
        .responsibility(new Responsibility("c2 resp1"))
        .build()
    ));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void oneClasseWithTwoResponsibilities() throws Exception {
    diagram = new Diagram(List.of(
      Class.builder()
        .name("className1")
        .responsibility(new Responsibility("c1 resp1"))
        .responsibility(new Responsibility("c1 resp2"))
        .build()
    ));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void twoClassesWithTwoResponsibilities() throws Exception {
    diagram = new Diagram(List.of(
      Class.builder()
        .name("className1")
        .responsibility(new Responsibility("c1 resp1"))
        .responsibility(new Responsibility("c1 resp2"))
        .build(),
      Class.builder()
        .name("className2")
        .responsibility(new Responsibility("c2 resp1"))
        .responsibility(new Responsibility("c2 resp2"))
        .build()
    ));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void generatingUnidirectionalConnections() throws Exception {
    diagram = new Diagram(List.of(
        Class.builder()
          .name("className1")
          .responsibility(new Responsibility("c1 resp1", "className2"))
          .responsibility(new Responsibility("c1 resp2", "className2"))
          .build(),
        Class.builder()
          .name("className2")
          .responsibility(new Responsibility("c2 resp1"))
          .responsibility(new Responsibility("c2 resp2", "className3"))
          .build(),
        Class.builder()
          .name("className3")
          .responsibility(new Responsibility("c3 resp1", "className1"))
          .responsibility(new Responsibility("c3 resp2"))
          .build()
        ));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  @Test
  void generatingBidirectionalConnections() throws Exception {
    diagram = new Diagram(List.of(
        Class.builder()
          .name("className1")
          .alias("a1")
          .responsibility(new Responsibility("c1 resp1", "className2"))
          .responsibility(new Responsibility("c1 resp2"))
          .build(),
        Class.builder()
          .name("className2")
          .responsibility(new Responsibility("c2 resp1"))
          .responsibility(new Responsibility("c2 resp2", "a1"))
          .build()
        ));
    generator = new DotGenerator(diagram, writer);

    generator.writeDiagram();

    Approvals.verify(writer.toString());
  }

  // TODO escaping text in class name, responsibility and collaborators
  /*
   * \\ --> \\
   * \: --> :
   * \n --> \l
   * " --> \"
   */
}
