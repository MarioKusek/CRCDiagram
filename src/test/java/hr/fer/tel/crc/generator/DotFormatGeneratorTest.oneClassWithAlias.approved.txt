digraph structs {
  node [shape=record];
  cl0 [label="{className1 [classAlias1] | {\l | \l}}"];

}
